pennylane/transforms/batch_input.py

```diff
@@ -73,25 +72,39 @@ def circuit(inputs, weights):
     array([0.46230079, 0.73971315, 0.95666004, 0.5355225 , 0.66180948,
             0.44519553, 0.93874261, 0.9483197 , 0.78737918, 0.90866411])>
     """
-    parameters = tape.get_parameters(trainable_only=False)
-
     argnum = tuple(argnum) if isinstance(argnum, (list, tuple)) else (int(argnum),)
 
-    non_trainable, trainable = [], []
-    for idx, param in enumerate(parameters):
-        if idx in argnum:
-            non_trainable.append(param)
-        else:
-            trainable.append(param)
+    trainable_params = tape.get_parameters()
+    all_parameters = sum(
+        (op.parameters if op.parameters != [] else [None] for op in tape.operations), []
+    )
+    argnum_params = [all_parameters[i] for i in argnum]
 
-    if len(np.unique([qml.math.shape(x)[0] for x in non_trainable])) != 1:
+    if any(
+        any(param is train_param for train_param in trainable_params) for param in argnum_params
+    ):
+        raise ValueError(
+            "Batched inputs must be non-trainable. Please make sure that the parameters indexed by "
+            + "'argnum' have 'requires_grad' set to False."
+        )
+
+    if len(np.unique([qml.math.shape(x)[0] for x in argnum_params])) != 1:
         raise ValueError(
             "Batch dimension for all gate arguments specified by 'argnum' must be the same."
         )
 
+    batch_size = qml.math.shape(argnum_params[0])[0]
```

Comment:
Could we unify these two steps?
```suggestion
    batch_dims = np.unique([qml.math.shape(x)[0] for x in argnum_params])
    if len(batch_dims) != 1:
        raise ValueError(
            "Batch dimension for all gate arguments specified by 'argnum' must be the same."
        )

    batch_size = batch_dims[0]
```

Reply:
Nice! 💯